En este archivo, voy a desarrollar el segundo sistema de recomendación que se me pide para este proyecto.
Este sistema de recomendación funciona de la siguiente manera:
![alt text](../img/user_item.jpg "Title")

Suponiendo que tenemos a un usuario 1 y a un usuario 2. Al usuario 1 le gusta el cubo de rubik, los dados y un juego de cartas. Al usuario 2 también le gusta el cubo de rubik y los dados, pero no ha jugado al juego de cartas.
 De forma muy simple, el sistema de recomendación user-item reconocería que estos dos usuarios comparten gustos similares, y al usuario 2 se le recomendaría el juego de cartas, que es jugado por otro usuario con gustos similares al suyo.  

El algoritmo esta fundamentado en que un usuario es similar a otro cuando le gustan (en este caso) los mismos juegos que a otro usuario. Entonces, si hay un juego que el usuario 2 no ha jugado, pero que el usuario 1 si, y son usuarios similares, es probable que al usuario 2 le guste el juego. 

Comenzamos importando las librerías necesarias para trabajar

In [1]:
# Para procesar los datos
import pandas as pd
import numpy as np
import scipy.stats as stats

# Para visualizar los datos
import seaborn as sns

# Para la similitud
from sklearn.metrics.pairwise import cosine_similarity

In [2]:
juegos = pd.read_parquet("../Datasets/steam_games_complete.parquet")
juegos.head()

item_id                item_name            developer  \
88310  761140      Lost Summoner Kitty            Kotoshiro   
88311  643980                Ironbound     Secret Level SRL   
88312  670290  Real Pool 3D - Poolians         Poolians.com   
88313  767400                  弹炸人2222                 彼岸领域   
88315  772540    Battle Royale Trainer  Trickjump Games Ltd   

                                                  genres  \
88310      [Action, Casual, Indie, Simulation, Strategy]   
88311               [Free to Play, Indie, RPG, Strategy]   
88312  [Casual, Free to Play, Indie, Simulation, Sports]   
88313                        [Action, Adventure, Casual]   
88315                    [Action, Adventure, Simulation]   

                                                    tags  \
88310      [Strategy, Action, Indie, Casual, Simulation]   
88311  [Free to Play, Strategy, Indie, RPG, Card Game...   
88312  [Free to Play, Simulation, Sports, Casual, Ind...   
88313                        [Action, Adventure, Casual]   
88315  [Action, Adventure, Simulation, FPS, Shooter, ...   

                                                   specs release_date  price  
88310                                    [Single-player]   2018-01-04   4.99  
88311  [Single-player, Multi-player, Online Multi-Pla...   2018-01-04   0.00  
88312  [Single-player, Multi-player, Online Multi-Pla...   2017-07-24   0.00  
88313                                    [Single-player]   2017-12-07   0.99  
88315                [Single-player, Steam Achievements]   2018-01-04   3.99

In [3]:
reseñas = pd.read_parquet("../Datasets/reviews_con_puntaje.parquet")
reseñas.head()

funny                     posted last_edited item_id  \
0         Posted November 5, 2011.                1250   
1            Posted July 15, 2011.               22200   
2           Posted April 21, 2011.               43110   
3            Posted June 24, 2014.              251610   
4        Posted September 8, 2013.              227300   

                                           helpful  recommend  \
0                                   No ratings yet       True   
1                                   No ratings yet       True   
2                                   No ratings yet       True   
3  15 of 20 people (75%) found this review helpful       True   
4     0 of 1 people (0%) found this review helpful       True   

                                              review            user_id  \
0  Simple yet with great replayability. In my opi...  76561197970982479   
1               It's unique and worth a playthrough.  76561197970982479   
2  Great atmosphere. The gunplay can be a bit chu...  76561197970982479   
3  I know what you think when you see this title ...            js41637   
4  For a simple (it's actually not all that simpl...            js41637   

   puntaje  
0        2  
1        2  
2        2  
3        2  
4        2

In [4]:
user_items = pd.read_parquet("../EDA/user_items_complete.parquet")
user_items.head()

item_id                  item_name  playtime_forever  playtime_2weeks  \
0      10             Counter-Strike                 6                0   
1      20      Team Fortress Classic                 0                0   
2      30              Day of Defeat                 7                0   
3      40         Deathmatch Classic                 0                0   
4      50  Half-Life: Opposing Force                 0                0   

             user_id  
0  76561197970982479  
1  76561197970982479  
2  76561197970982479  
3  76561197970982479  
4  76561197970982479

A continuación, veamos un poco la longitud de estos 3 conjuntos de datos.

In [5]:
print(len(juegos))
print()
print(len(reseñas))
print()
print(len(user_items))


22530

59305

5153209


Comencemos filtrando el dataset de user_items, ya que es el más largo de todos.

In [6]:
# Primero, observemos el valor máximo de la columna playtime_forever de user_items
user_items["playtime_forever"].max()

642773

Según deduzco, la unidad de tiempo en la que se midió este valor es en minutos, ya que (si fuesen horas) 642773 / 24 = 26782 (días jugados), 26782 / 365 = 73 (años jugados). Es una barbaridad, y es lógico pensar que esa no es la unidad de tiempo. En cambio, si consideramos que son minutos: 642773 / 60 = 10712 (horas jugadas), 10712 / 24 = 446 (días jugados), 446 / 365 = un año y meses. Tiene más sentido.


Voy a considerar a los usuarios que unicamente invirtieron historicamente más de 10 días (240 horas, 14.400 minutos) en el juego dado. Veamos cómo se reduce el dataset.

In [7]:
len(user_items[user_items["playtime_forever"] > 14400])

66487

Con eso pasaría de 5 millones de filas a sólo 66487. Las cuáles siguen siendo bastante, pero con esa consideración logré reducir el dataset en un 98% de su tamaño original.

In [8]:
# Guardamos los cambios en el dataset
user_items = user_items[user_items["playtime_forever"] > 14400]
user_items

item_id                         item_name  playtime_forever  \
178         730  Counter-Strike: Global Offensive             23532   
1301     466170           Idling to Rule the Gods             28545   
1354      12210               Grand Theft Auto IV             52062   
1361      21660                 Street Fighter IV             23903   
1907        730  Counter-Strike: Global Offensive             19800   
...         ...                               ...               ...   
5150878  304930                          Unturned             14476   
5150968     730  Counter-Strike: Global Offensive             29789   
5151088     730  Counter-Strike: Global Offensive             16635   
5152561  304930                          Unturned             20201   
5152587     730  Counter-Strike: Global Offensive             20016   

         playtime_2weeks            user_id  
178                    0  76561197970982479  
1301                1554          evcentric  
1354                   0         Riot-Punch  
1361                   0         Riot-Punch  
1907                   0              doctr  
...                  ...                ...  
5150878                0  76561198285507552  
5150968                0            sexyawp  
5151088                0            tarik22  
5152561                0  76561198304604920  
5152587               96  76561198310819422  

[66487 rows x 5 columns]

In [9]:
len(user_items["item_id"].unique())

1071

Lo que voy a hacer a continuación es tener en cuenta sólo aquellos juegos que tienen más de 100 reseñas.

In [10]:
juegos_con_mas_de_100_reseñas = reseñas["item_id"].value_counts() > 100
# juegos_con_mas_de_100_reseñas
juegos_con_mas_de_100_reseñas.value_counts()
# reseñas[reseñas["item_id"] == (reseñas["item_id"].value_counts() > 100)]

count
False    3583
True       99
Name: count, dtype: int64

In [11]:
len(reseñas[reseñas["item_id"] == "211820"])

351

In [12]:
reseñas

funny                     posted  \
0                                         Posted November 5, 2011.   
1                                            Posted July 15, 2011.   
2                                           Posted April 21, 2011.   
3                                            Posted June 24, 2014.   
4                                        Posted September 8, 2013.   
...                                 ...                        ...   
59300                                              Posted July 10.   
59301                                               Posted July 8.   
59302  1 person found this review funny             Posted July 3.   
59303                                              Posted July 20.   
59304                                               Posted July 2.   

      last_edited item_id                                          helpful  \
0                    1250                                   No ratings yet   
1                   22200                                   No ratings yet   
2                   43110                                   No ratings yet   
3                  251610  15 of 20 people (75%) found this review helpful   
4                  227300     0 of 1 people (0%) found this review helpful   
...           ...     ...                                              ...   
59300                  70                                   No ratings yet   
59301              362890                                   No ratings yet   
59302              273110    1 of 2 people (50%) found this review helpful   
59303                 730                                   No ratings yet   
59304                 440                                   No ratings yet   

       recommend                                             review  \
0           True  Simple yet with great replayability. In my opi...   
1           True               It's unique and worth a playthrough.   
2           True  Great atmosphere. The gunplay can be a bit chu...   
3           True  I know what you think when you see this title ...   
4           True  For a simple (it's actually not all that simpl...   
...          ...                                                ...   
59300       True  a must have classic from steam definitely wort...   
59301       True  this game is a perfect remake of the original ...   
59302       True  had so much fun plaing this and collecting res...   
59303       True                                                 :D   
59304       True                                     so much fun :D   

                 user_id  puntaje  
0      76561197970982479        2  
1      76561197970982479        2  
2      76561197970982479        2  
3                js41637        2  
4                js41637        2  
...                  ...      ...  
59300  76561198312638244        2  
59301  76561198312638244        2  
59302        LydiaMorley        2  
59303        LydiaMorley        2  
59304        LydiaMorley        2  

[59305 rows x 9 columns]

In [13]:
lista_booleana_para_mascara = []
for id in reseñas["item_id"]:
    lista_booleana_para_mascara.append(juegos_con_mas_de_100_reseñas[id])


In [14]:
lista_booleana_para_mascara

[True,
 False,
 False,
 False,
 True,
 False,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 False,
 True,
 True,
 True,
 True,
 False,
 True,
 True,
 False,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 False,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 False,
 True,
 True,
 True,
 False,
 True,
 True,
 False,
 False,
 False,
 True,
 False,
 True,
 False,
 False,
 True,
 False,
 True,
 True,
 False,
 True,
 True,
 True,
 True,
 False,
 True,
 False,
 True,
 True,
 False,
 True,
 True,
 True,
 True,
 True,
 True,
 False,
 False,
 True,
 True,
 True,
 False,
 True,
 True,
 True,
 False,
 True,
 False,
 True,
 True,
 False,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 False,
 False,
 True,
 False,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 False,
 False,
 False,
 True,
 False,
 Fa

In [15]:
reseñas = reseñas[lista_booleana_para_mascara]
reseñas

funny                     posted  \
0                                         Posted November 5, 2011.   
4                                        Posted September 8, 2013.   
6                                               Posted February 3.   
16                                       Posted November 22, 2012.   
18     3 people found this review funny     Posted April 15, 2014.   
...                                 ...                        ...   
59295                                               Posted May 31.   
59296                                              Posted June 17.   
59297  1 person found this review funny            Posted June 23.   
59303                                              Posted July 20.   
59304                                               Posted July 2.   

      last_edited item_id                                          helpful  \
0                    1250                                   No ratings yet   
4                  227300     0 of 1 people (0%) found this review helpful   
6                  248820                                   No ratings yet   
16                 207610                                   No ratings yet   
18                 211420  35 of 43 people (81%) found this review helpful   
...           ...     ...                                              ...   
59295              261030     0 of 1 people (0%) found this review helpful   
59296                 730     0 of 1 people (0%) found this review helpful   
59297                 570   1 of 1 people (100%) found this review helpful   
59303                 730                                   No ratings yet   
59304                 440                                   No ratings yet   

       recommend                                             review  \
0           True  Simple yet with great replayability. In my opi...   
4           True  For a simple (it's actually not all that simpl...   
6           True  A suitably punishing roguelike platformer.  Wi...   
16          True  The ending to this game is.... ♥♥♥♥♥♥♥.... Jus...   
18          True                                            Git gud   
...          ...                                                ...   
59295       True  I cried in the end its so sadding ]'; I wish l...   
59296       True  Gra naprawdę fajna.Ale jest kilka rzeczy do kt...   
59297       True                                          Well Done   
59303       True                                                 :D   
59304       True                                     so much fun :D   

                 user_id  puntaje  
0      76561197970982479        2  
4                js41637        2  
6              evcentric        2  
16                 doctr        2  
18             maplemage        2  
...                  ...      ...  
59295  76561198306599751        0  
59296           Ghoustik        1  
59297  76561198310819422        2  
59303        LydiaMorley        2  
59304        LydiaMorley        2  

[33714 rows x 9 columns]

In [16]:
usuarios_con_mas_de_5_reseñas = reseñas["user_id"].value_counts() > 5
usuarios_con_mas_de_5_reseñas

user_id
76561198063316459     True
OfficialSenix         True
76561198094304449     True
ITCGaming101          True
ironcass              True
                     ...  
Melonman231          False
pandashan            False
76561198059167447    False
76561198116801566    False
76561197970982479    False
Name: count, Length: 19620, dtype: bool

In [17]:
usuarios_con_mas_de_5_reseñas.value_counts()

count
False    19164
True       456
Name: count, dtype: int64

In [18]:
lista_booleana_para_mascara = []
for id in reseñas["user_id"]:
    lista_booleana_para_mascara.append(usuarios_con_mas_de_5_reseñas[id])

lista_booleana_para_mascara

[False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 True,
 True,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 True,
 True,
 True,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 Fals

In [19]:
reseñas[lista_booleana_para_mascara]

funny                    posted last_edited  \
39                                          Posted July 14, 2014.               
40                                         Posted April 27, 2013.               
41     1 person found this review funny     Posted July 20, 2015.               
42                                       Posted November 4, 2013.               
43                                          Posted July 12, 2013.               
...                                 ...                       ...         ...   
58954  2 people found this review funny      Posted June 1, 2015.               
58955  2 people found this review funny      Posted June 1, 2015.               
58956                                         Posted May 7, 2015.               
58957  5 people found this review funny      Posted May 18, 2015.               
58958                                        Posted May 14, 2015.               

      item_id                                         helpful  recommend  \
39     304930   2 of 4 people (50%) found this review helpful       True   
40        570    0 of 1 people (0%) found this review helpful       True   
41     218620   2 of 6 people (33%) found this review helpful       True   
42     224260                                  No ratings yet       True   
43       1250                                  No ratings yet       True   
...       ...                                             ...        ...   
58954  230410   6 of 9 people (67%) found this review helpful       True   
58955  304930  6 of 11 people (55%) found this review helpful       True   
58956     730   1 of 2 people (50%) found this review helpful       True   
58957  236390  8 of 17 people (47%) found this review helpful       True   
58958     570                                  No ratings yet       True   

                                                  review          user_id  \
39     At first, it looked to me that..."Wow, just an...         DJKamBer   
40     Awesome graphics, texture. Reqiures HARDCORE t...         DJKamBer   
41     Upside to PAYDAY 2:+ Updates are released ever...         DJKamBer   
42     Awesome game! It's like L4D2 and Killing Floor...         DJKamBer   
43     Compared to Left 4 Dead 2, this game REALLY gi...         DJKamBer   
...                                                  ...              ...   
58954             A game for in between. Nothing special  SponsoredByBenQ   
58955                                          Nice Game  SponsoredByBenQ   
58956  top game. mm sucks sometimes cause of many rus...  SponsoredByBenQ   
58957  A game for in between. Despite all this, a goo...  SponsoredByBenQ   
58958              Pretty nice game, but I prefer LoL :/  SponsoredByBenQ   

       puntaje  
39           2  
40           2  
41           2  
42           2  
43           2  
...        ...  
58954        1  
58955        2  
58956        2  
58957        2  
58958        2  

[3089 rows x 9 columns]

In [20]:
# Se supone que este usuario DJKamBer tiene más de 5 reseñas, comprobamos:
print(usuarios_con_mas_de_5_reseñas["DJKamBer"])
print(len(reseñas[reseñas["user_id"] == "DJKamBer"]))

True
6


Aplico la mascara para dejar unicamente a los 456 usuarios con más de 5 reseñas

In [21]:
reseñas = reseñas[lista_booleana_para_mascara]
reseñas

funny                    posted last_edited  \
39                                          Posted July 14, 2014.               
40                                         Posted April 27, 2013.               
41     1 person found this review funny     Posted July 20, 2015.               
42                                       Posted November 4, 2013.               
43                                          Posted July 12, 2013.               
...                                 ...                       ...         ...   
58954  2 people found this review funny      Posted June 1, 2015.               
58955  2 people found this review funny      Posted June 1, 2015.               
58956                                         Posted May 7, 2015.               
58957  5 people found this review funny      Posted May 18, 2015.               
58958                                        Posted May 14, 2015.               

      item_id                                         helpful  recommend  \
39     304930   2 of 4 people (50%) found this review helpful       True   
40        570    0 of 1 people (0%) found this review helpful       True   
41     218620   2 of 6 people (33%) found this review helpful       True   
42     224260                                  No ratings yet       True   
43       1250                                  No ratings yet       True   
...       ...                                             ...        ...   
58954  230410   6 of 9 people (67%) found this review helpful       True   
58955  304930  6 of 11 people (55%) found this review helpful       True   
58956     730   1 of 2 people (50%) found this review helpful       True   
58957  236390  8 of 17 people (47%) found this review helpful       True   
58958     570                                  No ratings yet       True   

                                                  review          user_id  \
39     At first, it looked to me that..."Wow, just an...         DJKamBer   
40     Awesome graphics, texture. Reqiures HARDCORE t...         DJKamBer   
41     Upside to PAYDAY 2:+ Updates are released ever...         DJKamBer   
42     Awesome game! It's like L4D2 and Killing Floor...         DJKamBer   
43     Compared to Left 4 Dead 2, this game REALLY gi...         DJKamBer   
...                                                  ...              ...   
58954             A game for in between. Nothing special  SponsoredByBenQ   
58955                                          Nice Game  SponsoredByBenQ   
58956  top game. mm sucks sometimes cause of many rus...  SponsoredByBenQ   
58957  A game for in between. Despite all this, a goo...  SponsoredByBenQ   
58958              Pretty nice game, but I prefer LoL :/  SponsoredByBenQ   

       puntaje  
39           2  
40           2  
41           2  
42           2  
43           2  
...        ...  
58954        1  
58955        2  
58956        2  
58957        2  
58958        2  

[3089 rows x 9 columns]

In [43]:
# Lo que voy a hacer es aumentar el valor del puntaje de las reseñas en 1, considerando las negativas con un 1, las neutras con un 2 y las positivas con un tres.
reseñas["puntaje"] += 1
reseñas

C:\Users\crisr\AppData\Local\Temp\ipykernel_20148\1416905613.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  reseñas["puntaje"] += 1


funny                    posted last_edited  \
39                                          Posted July 14, 2014.               
40                                         Posted April 27, 2013.               
41     1 person found this review funny     Posted July 20, 2015.               
42                                       Posted November 4, 2013.               
43                                          Posted July 12, 2013.               
...                                 ...                       ...         ...   
58954  2 people found this review funny      Posted June 1, 2015.               
58955  2 people found this review funny      Posted June 1, 2015.               
58956                                         Posted May 7, 2015.               
58957  5 people found this review funny      Posted May 18, 2015.               
58958                                        Posted May 14, 2015.               

      item_id                                         helpful  recommend  \
39     304930   2 of 4 people (50%) found this review helpful       True   
40        570    0 of 1 people (0%) found this review helpful       True   
41     218620   2 of 6 people (33%) found this review helpful       True   
42     224260                                  No ratings yet       True   
43       1250                                  No ratings yet       True   
...       ...                                             ...        ...   
58954  230410   6 of 9 people (67%) found this review helpful       True   
58955  304930  6 of 11 people (55%) found this review helpful       True   
58956     730   1 of 2 people (50%) found this review helpful       True   
58957  236390  8 of 17 people (47%) found this review helpful       True   
58958     570                                  No ratings yet       True   

                                                  review          user_id  \
39     At first, it looked to me that..."Wow, just an...         DJKamBer   
40     Awesome graphics, texture. Reqiures HARDCORE t...         DJKamBer   
41     Upside to PAYDAY 2:+ Updates are released ever...         DJKamBer   
42     Awesome game! It's like L4D2 and Killing Floor...         DJKamBer   
43     Compared to Left 4 Dead 2, this game REALLY gi...         DJKamBer   
...                                                  ...              ...   
58954             A game for in between. Nothing special  SponsoredByBenQ   
58955                                          Nice Game  SponsoredByBenQ   
58956  top game. mm sucks sometimes cause of many rus...  SponsoredByBenQ   
58957  A game for in between. Despite all this, a goo...  SponsoredByBenQ   
58958              Pretty nice game, but I prefer LoL :/  SponsoredByBenQ   

       puntaje  
39           4  
40           4  
41           4  
42           4  
43           4  
...        ...  
58954        3  
58955        4  
58956        4  
58957        4  
58958        4  

[3089 rows x 9 columns]

In [23]:
print(len(reseñas["item_id"].unique())) # 99 juegos
print(len(reseñas["user_id"].unique())) # 456 usuarios

99
456


Ahora lo que voy a hacer es unir las tablas de reseña y de juegos para tener disponible el nombre del juego, y no sólo el item_id, lo cuál me va a simplificar mucho más adelante

In [24]:
juegos_y_reseñas = pd.merge(reseñas,juegos,on="item_id")
juegos_y_reseñas

funny                    posted last_edited  \
0                                          Posted July 14, 2014.               
1                                       Posted November 4, 2013.               
2                                          Posted July 12, 2013.               
3                                        Posted August 19, 2012.               
4                                                Posted July 30.               
...                                ...                       ...         ...   
2551  2 people found this review funny      Posted June 1, 2015.               
2552                                        Posted May 18, 2015.               
2553  2 people found this review funny      Posted June 1, 2015.               
2554  2 people found this review funny      Posted June 1, 2015.               
2555                                         Posted May 7, 2015.               

     item_id                                          helpful  recommend  \
0     304930    2 of 4 people (50%) found this review helpful       True   
1     224260                                   No ratings yet       True   
2       1250                                   No ratings yet       True   
3        440                                   No ratings yet       True   
4       4000     0 of 1 people (0%) found this review helpful       True   
...      ...                                              ...        ...   
2551  107410  23 of 26 people (88%) found this review helpful       True   
2552     440    5 of 6 people (83%) found this review helpful       True   
2553  230410    6 of 9 people (67%) found this review helpful       True   
2554  304930   6 of 11 people (55%) found this review helpful       True   
2555     730    1 of 2 people (50%) found this review helpful       True   

                                                 review          user_id  \
0     At first, it looked to me that..."Wow, just an...         DJKamBer   
1     Awesome game! It's like L4D2 and Killing Floor...         DJKamBer   
2     Compared to Left 4 Dead 2, this game REALLY gi...         DJKamBer   
3                          You won't regret playing it!         DJKamBer   
4     Vale a pena a pagar 20 R$ nesse jogo porque:  ...        diego9031   
...                                                 ...              ...   
2551  A really nice game. If you want to play a real...  SponsoredByBenQ   
2552  A game for in between. Despite all this, a goo...  SponsoredByBenQ   
2553             A game for in between. Nothing special  SponsoredByBenQ   
2554                                          Nice Game  SponsoredByBenQ   
2555  top game. mm sucks sometimes cause of many rus...  SponsoredByBenQ   

      puntaje                         item_name                  developer  \
0           3                          Unturned      Smartly Dressed Games   
1           3              No More Room in Hell  No More Room in Hell Team   
2           3                     Killing Floor       Tripwire Interactive   
3           3                   Team Fortress 2                      Valve   
4           2                       Garry's Mod          Facepunch Studios   
...       ...                               ...                        ...   
2551        3                            Arma 3        Bohemia Interactive   
2552        3                   Team Fortress 2                      Valve   
2553        2                          Warframe           Digital Extremes   
2554        3                          Unturned      Smartly Dressed Games   
2555        3  Counter-Strike: Global Offensive                      Valve   

                                                genres  \
0     [Action, Adventure, Casual, Free to Play, Indie]   
1                        [Action, Free to Play, Indie]   
2                                             [Action]   
3                               [Action, Free to Play]   
4                  

A continuación elimino las columnas y filas innecesarias para una correcta visualización y óptimo procesamiento

In [25]:
# Empecemos por las columnas
juegos_y_reseñas.columns 

Index(['funny', 'posted', 'last_edited', 'item_id', 'helpful', 'recommend',
       'review', 'user_id', 'puntaje', 'item_name', 'developer', 'genres',
       'tags', 'specs', 'release_date', 'price'],
      dtype='object')

Las únicas columnas que me sirven son:

- 'user_id'
- 'puntaje'
- 'item_name'

User ID me sirve para usarlo de índice en la tabla que voy a crear a continuación, item_name me sirve para usarlo de nombre de columnas, y puntajes para los valores dentro de esa tabla.


Procedo a eliminar las columnas que no me sirven

In [26]:
for columna in juegos_y_reseñas.columns:
    if columna not in ["user_id","puntaje","item_name"]:
        juegos_y_reseñas.drop(columna, inplace=True, axis=1)
    
print(juegos_y_reseñas.columns)

Index(['user_id', 'puntaje', 'item_name'], dtype='object')


Ahora voy con las filas

In [27]:
juegos_y_reseñas

user_id  puntaje                         item_name
0            DJKamBer        3                          Unturned
1            DJKamBer        3              No More Room in Hell
2            DJKamBer        3                     Killing Floor
3            DJKamBer        3                   Team Fortress 2
4           diego9031        2                       Garry's Mod
...               ...      ...                               ...
2551  SponsoredByBenQ        3                            Arma 3
2552  SponsoredByBenQ        3                   Team Fortress 2
2553  SponsoredByBenQ        2                          Warframe
2554  SponsoredByBenQ        3                          Unturned
2555  SponsoredByBenQ        3  Counter-Strike: Global Offensive

[2556 rows x 3 columns]

In [28]:
# Chequeo de nulos
juegos_y_reseñas.isnull().sum() # No hay, entonces no elimino ninguna fila

user_id      0
puntaje      0
item_name    0
dtype: int64

Procedo a crear la tabla

In [29]:
matrix = juegos_y_reseñas.pivot_table(index='user_id',columns='item_name',values='puntaje')
matrix.head()

item_name                  APB Reloaded  ARK: Survival Evolved  \
user_id                                                          
09879655452567                      NaN                    3.0   
10051997                            NaN                    NaN   
1011001                             NaN                    NaN   
111222333444555666888               NaN                    NaN   
1234567io9872345678765432           1.0                    NaN   

item_name                  Ace of Spades: Battle Builder  \
user_id                                                    
09879655452567                                       NaN   
10051997                                             NaN   
1011001                                              NaN   
111222333444555666888                                NaN   
1234567io9872345678765432                            NaN   

item_name                  AdVenture Capitalist  Arma 3  \
user_id                                                   
09879655452567                              NaN     NaN   
10051997                                    NaN     NaN   
1011001                                     NaN     NaN   
111222333444555666888                       NaN     NaN   
1234567io9872345678765432                   NaN     1.0   

item_name                  Awesomenauts - the 2D moba  \
user_id                                                 
09879655452567                                    NaN   
10051997                                          NaN   
1011001                                           NaN   
111222333444555666888                             NaN   
1234567io9872345678765432                         NaN   

item_name                  Bad Rats: the Rats' Revenge  BattleBlock Theater®  \
user_id                                                                        
09879655452567                                     NaN                   NaN   
10051997                                           NaN                   NaN   
1011001                                            NaN                   NaN   
111222333444555666888                              NaN                   1.0   
1234567io9872345678765432                          NaN                   NaN   

item_name                  BioShock Infinite  Blacklight: Retribution  ...  \
user_id                                                                ...   
09879655452567                           NaN                      NaN  ...   
10051997                                 NaN                      NaN  ...   
1011001                                  NaN                      NaN  ...   
111222333444555666888                    2.0                      NaN  ...   
1234567io9872345678765432                NaN                      NaN  ...   

item_name                  The Stanley Parable  The Walking Dead  \
user_id                                                            
09879655452567                             NaN               NaN   
10051997                                   NaN               NaN   
1011001                                    NaN               NaN   
111222333444555666888                      NaN               NaN   
1234567io9872345678765432                  NaN               NaN   

item_name                  The Walking Dead: Season 2  Tomb Raider  Trove  \
user_id                                                                     
09879655452567                                    NaN          NaN    1.0   
10051997                                          NaN          NaN    NaN   
1011001                                           NaN          NaN    NaN   
111222333444555666888                             NaN          NaN    NaN   
1234567io9872345678765432                         NaN          NaN    NaN   

item_name                  Undertale  Unturned  Verdun  Warframe  \
user_id                                                            
09879655452567                   NaN     

Ahora le resto a cada fila la media de la fila para normalizar la tabla

In [30]:
matrix_norm = matrix.subtract(matrix.mean(axis=1), axis = 'rows')
matrix_norm.head()

item_name                  APB Reloaded  ARK: Survival Evolved  \
user_id                                                          
09879655452567                      NaN                    1.0   
10051997                            NaN                    NaN   
1011001                             NaN                    NaN   
111222333444555666888               NaN                    NaN   
1234567io9872345678765432          -0.2                    NaN   

item_name                  Ace of Spades: Battle Builder  \
user_id                                                    
09879655452567                                       NaN   
10051997                                             NaN   
1011001                                              NaN   
111222333444555666888                                NaN   
1234567io9872345678765432                            NaN   

item_name                  AdVenture Capitalist  Arma 3  \
user_id                                                   
09879655452567                              NaN     NaN   
10051997                                    NaN     NaN   
1011001                                     NaN     NaN   
111222333444555666888                       NaN     NaN   
1234567io9872345678765432                   NaN    -0.2   

item_name                  Awesomenauts - the 2D moba  \
user_id                                                 
09879655452567                                    NaN   
10051997                                          NaN   
1011001                                           NaN   
111222333444555666888                             NaN   
1234567io9872345678765432                         NaN   

item_name                  Bad Rats: the Rats' Revenge  BattleBlock Theater®  \
user_id                                                                        
09879655452567                                     NaN                   NaN   
10051997                                           NaN                   NaN   
1011001                                            NaN                   NaN   
111222333444555666888                              NaN             -0.666667   
1234567io9872345678765432                          NaN                   NaN   

item_name                  BioShock Infinite  Blacklight: Retribution  ...  \
user_id                                                                ...   
09879655452567                           NaN                      NaN  ...   
10051997                                 NaN                      NaN  ...   
1011001                                  NaN                      NaN  ...   
111222333444555666888               0.333333                      NaN  ...   
1234567io9872345678765432                NaN                      NaN  ...   

item_name                  The Stanley Parable  The Walking Dead  \
user_id                                                            
09879655452567                             NaN               NaN   
10051997                                   NaN               NaN   
1011001                                    NaN               NaN   
111222333444555666888                      NaN               NaN   
1234567io9872345678765432                  NaN               NaN   

item_name                  The Walking Dead: Season 2  Tomb Raider  Trove  \
user_id                                                                     
09879655452567                                    NaN          NaN   -1.0   
10051997                                          NaN          NaN    NaN   
1011001                                           NaN          NaN    NaN   
111222333444555666888                             NaN          NaN    NaN   
1234567io9872345678765432                         NaN          NaN    NaN   

item_name                  Undertale  Unturned  Verdun  Warframe  \
user_id                                                            
09879655452567                   NaN     

In [31]:
matrix_norm

item_name                  APB Reloaded  ARK: Survival Evolved  \
user_id                                                          
09879655452567                      NaN                    1.0   
10051997                            NaN                    NaN   
1011001                             NaN                    NaN   
111222333444555666888               NaN                    NaN   
1234567io9872345678765432          -0.2                    NaN   
...                                 ...                    ...   
zakbot                              NaN                   -0.6   
zayyntt                             NaN                    NaN   
zerzang                             NaN                    NaN   
zrustz16                            NaN                    NaN   
zyr0n1c                             NaN                    NaN   

item_name                  Ace of Spades: Battle Builder  \
user_id                                                    
09879655452567                                       NaN   
10051997                                             NaN   
1011001                                              NaN   
111222333444555666888                                NaN   
1234567io9872345678765432                            NaN   
...                                                  ...   
zakbot                                              -0.6   
zayyntt                                              NaN   
zerzang                                              NaN   
zrustz16                                             NaN   
zyr0n1c                                              NaN   

item_name                  AdVenture Capitalist  Arma 3  \
user_id                                                   
09879655452567                              NaN     NaN   
10051997                                    NaN     NaN   
1011001                                     NaN     NaN   
111222333444555666888                       NaN     NaN   
1234567io9872345678765432                   NaN    -0.2   
...                                         ...     ...   
zakbot                                      NaN     NaN   
zayyntt                                     NaN     NaN   
zerzang                                     NaN     NaN   
zrustz16                                    NaN     NaN   
zyr0n1c                                     NaN     NaN   

item_name                  Awesomenauts - the 2D moba  \
user_id                                                 
09879655452567                                    NaN   
10051997                                          NaN   
1011001                                           NaN   
111222333444555666888                             NaN   
1234567io9872345678765432                         NaN   
...                                               ...   
zakbot                                            NaN   
zayyntt                                           NaN   
zerzang                                           NaN   
zrustz16                                          NaN   
zyr0n1c                                           NaN   

item_name                  Bad Rats: the Rats' Revenge  BattleBlock Theater®  \
user_id                                                                        
09879655452567                                     NaN                   NaN   
10051997                                           NaN                   NaN   
1011001                                            NaN                   NaN   
111222333444555666888                              NaN             -0.666667   
1234567io9872345678765432                          NaN                   NaN   
...                                                ...                   ...   
zakbot                                             NaN                   NaN   
zayyntt                                            NaN                   NaN   
zerzang                                           

In [32]:
# Rellenar NaN con 0s ya que el coseno no se ve afectado por ceros
matrix_norm_filled = matrix_norm.fillna(0)

# Calcular la similitud del coseno
similitud_del_coseno = cosine_similarity(matrix_norm_filled)

# Convertir a DataFrame para mejor legibilidad
similitud_del_coseno

array([[ 1.        ,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       [ 0.        ,  1.        , -0.15474612, ...,  0.        ,
        -0.59628479,  0.        ],
       [ 0.        , -0.15474612,  1.        , ...,  0.        ,
         0.2076137 ,  0.        ],
       ...,
       [ 0.        ,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       [ 0.        , -0.59628479,  0.2076137 , ...,  0.        ,
         1.        ,  0.        ],
       [ 0.        ,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ]])

Ahora lo que voy a hacer es ver la similitud entre los usuarios, colocandolos tanto en el indice cómo en las columnas

In [33]:
# Convertir a DataFrame para mejor legibilidad
similitud_de_usuarios_coseno = pd.DataFrame(similitud_del_coseno, index=matrix_norm.index, columns=matrix_norm.index)

similitud_de_usuarios_coseno

user_id                    09879655452567  10051997   1011001  \
user_id                                                         
09879655452567                   1.000000  0.000000  0.000000   
10051997                         0.000000  1.000000 -0.154746   
1011001                          0.000000 -0.154746  1.000000   
111222333444555666888            0.000000 -0.333333  0.154746   
1234567io9872345678765432        0.000000 -0.730297  0.355983   
...                                   ...       ...       ...   
zakbot                          -0.167705  0.000000  0.000000   
zayyntt                          0.000000 -0.091287 -0.084758   
zerzang                          0.000000  0.000000  0.000000   
zrustz16                         0.000000 -0.596285  0.207614   
zyr0n1c                          0.000000  0.000000  0.000000   

user_id                    111222333444555666888  1234567io9872345678765432  \
user_id                                                                       
09879655452567                          0.000000                   0.000000   
10051997                               -0.333333                  -0.730297   
1011001                                 0.154746                   0.355983   
111222333444555666888                   1.000000                   0.365148   
1234567io9872345678765432               0.365148                   1.000000   
...                                          ...                        ...   
zakbot                                  0.000000                   0.000000   
zayyntt                                 0.000000                   0.050000   
zerzang                                 0.000000                   0.000000   
zrustz16                                0.223607                   0.571548   
zyr0n1c                                 0.000000                   0.000000   

user_id                    1873410337     1snap    210396     29123      2sBs  \
user_id                                                                         
09879655452567               0.111803  0.265334  0.000000  0.000000  0.000000   
10051997                     0.821584  0.000000  0.117851  0.288675  0.484481   
1011001                     -0.339032  0.000000  0.109422  0.268028 -0.034602   
111222333444555666888       -0.365148  0.000000  0.000000  0.000000 -0.186339   
1234567io9872345678765432   -0.800000  0.000000  0.000000  0.000000 -0.306186   
...                               ...       ...       ...       ...       ...   
zakbot                      -0.075000  0.000000  0.000000  0.000000  0.000000   
zayyntt                      0.000000  0.339032 -0.064550 -0.158114  0.183712   
zerzang                      0.000000  0.000000  0.000000  0.000000  0.000000   
zrustz16                    -0.571548  0.000000  0.000000  0.000000 -0.250000   
zyr0n1c                      0.000000  0.000000  0.000000  0.000000  0.000000   

user_id                    ...  xtomx_freedom    yotuic  you_re_ded  \
user_id                    ...                                        
09879655452567             ...       0.000000  0.000000    0.000000   
10051997                   ...       0.521749 -0.125988    0.195180   
1011001                    ...       0.069205  0.058489    0.181220   
111222333444555666888      ...      -0.149071  0.062994    0.000000   
1234567io9872345678765432  ...      -0.244949  0.345033   -0.106904   
...                        ...            ...       ...         ...   
zakbot                     ...       0.000000  0.000000    0.000000   
zayyntt                    ...      -0.122474  0.207020   -0.213809   
zerzang                    ...       0.000000  0.000000    0.000000   
zrustz16                   ...      -0.200000  0.422577   -0.174574   
zyr0n1c                    ...       0.000000  0.000000    0.000000   

user_id                    youngbenaffleck   zaaikbr    zakbot   zayyntt  \
user_id                                                                 

Acá comenzaría el primer prototipo del sistema de recomendación

In [34]:
# Elijo un usuario cualquiera para encontrar similitudes, en mi caso (el primero):
usuario_elegido = "09879655452567"	

# Elimino la fila con el usuario elegido y muestro el dataframe
similitud_de_usuarios_coseno.drop(index=usuario_elegido,inplace=True)
similitud_de_usuarios_coseno.head()

user_id                    09879655452567  10051997   1011001  \
user_id                                                         
10051997                         0.000000  1.000000 -0.154746   
1011001                          0.000000 -0.154746  1.000000   
111222333444555666888            0.000000 -0.333333  0.154746   
1234567io9872345678765432        0.000000 -0.730297  0.355983   
1873410337                       0.111803  0.821584 -0.339032   

user_id                    111222333444555666888  1234567io9872345678765432  \
user_id                                                                       
10051997                               -0.333333                  -0.730297   
1011001                                 0.154746                   0.355983   
111222333444555666888                   1.000000                   0.365148   
1234567io9872345678765432               0.365148                   1.000000   
1873410337                             -0.365148                  -0.800000   

user_id                    1873410337  1snap    210396     29123      2sBs  \
user_id                                                                      
10051997                     0.821584    0.0  0.117851  0.288675  0.484481   
1011001                     -0.339032    0.0  0.109422  0.268028 -0.034602   
111222333444555666888       -0.365148    0.0  0.000000  0.000000 -0.186339   
1234567io9872345678765432   -0.800000    0.0  0.000000  0.000000 -0.306186   
1873410337                   1.000000    0.0  0.000000  0.000000  0.408248   

user_id                    ...  xtomx_freedom    yotuic  you_re_ded  \
user_id                    ...                                        
10051997                   ...       0.521749 -0.125988    0.195180   
1011001                    ...       0.069205  0.058489    0.181220   
111222333444555666888      ...      -0.149071  0.062994    0.000000   
1234567io9872345678765432  ...      -0.244949  0.345033   -0.106904   
1873410337                 ...       0.244949 -0.138013    0.000000   

user_id                    youngbenaffleck   zaaikbr  zakbot   zayyntt  \
user_id                                                                  
10051997                          0.182574  0.000000   0.000 -0.091287   
1011001                          -0.288177  0.346023   0.000 -0.084758   
111222333444555666888            -0.091287  0.000000   0.000  0.000000   
1234567io9872345678765432        -0.275000  0.000000   0.000  0.050000   
1873410337                        0.275000  0.000000  -0.075  0.000000   

user_id                    zerzang  zrustz16  zyr0n1c  
user_id                                                
10051997                       0.0 -0.596285      0.0  
1011001                        0.0  0.207614      0.0  
111222333444555666888          0.0  0.223607      0.0  
1234567io9872345678765432      0.0  0.571548      0.0  
1873410337                     0.0 -0.571548      0.0  

[5 rows x 456 columns]

In [35]:
# Elijo una determinada cantidad de usuarios similares para encontrar, en mi caso voy a hacer el top 9 de usuarios más similares
cantidad_de_usuarios_similares = 9

# El porcentaje minimo de similitud lo establecí en un 36%
similitud_minima_entre_usuarios = 0.36

In [36]:
# Busco los usuarios similares y los ordeno desde el más similar al menos similar
usuarios_similares = similitud_de_usuarios_coseno[similitud_de_usuarios_coseno[usuario_elegido] > similitud_minima_entre_usuarios][usuario_elegido].sort_values(ascending=False)[:cantidad_de_usuarios_similares]

In [37]:
print(f'Los usuarios similares al usuario {usuario_elegido} son', usuarios_similares)

Los usuarios similares al usuario 09879655452567 son user_id
tarjla               0.462910
ImWinKo              0.433013
WCM03                0.381385
76561198096443555    0.365148
ShevilleWarhand      0.365148
Name: 09879655452567, dtype: float64


Vemos que en el caso del usuario elegido, el usuario que "mas se parece" tiene una similitud de un 46%

A continuación, voy a eliminar todos los juegos excepto aquellos que jugó el usuario elegido.

In [38]:
juegos_jugados_por_usuario_elegido = matrix_norm[matrix_norm.index == usuario_elegido].dropna(axis=1,how="all")
juegos_jugados_por_usuario_elegido


item_name       ARK: Survival Evolved  Garry's Mod  Heroes &amp; Generals  \
user_id                                                                     
09879655452567                    1.0          0.0                    1.0   

item_name       The Forest  Trove  
user_id                            
09879655452567        -1.0   -1.0

In [39]:
# Juegos que usuarios similares han jugado. Eliminamos los juegos que ninguno de los usuarios similares haya jugado
juegos_de_usuarios_similares = matrix_norm[matrix_norm.index.isin(usuarios_similares.index)].dropna(axis=1, how='all')
juegos_de_usuarios_similares

item_name            Arma 3  Call of Duty®: Black Ops III  \
user_id                                                     
76561198096443555       NaN                           NaN   
ImWinKo            0.666667                           NaN   
ShevilleWarhand         NaN                      0.333333   
WCM03                   NaN                           NaN   
tarjla                  NaN                           NaN   

item_name          Counter-Strike: Global Offensive  Counter-Strike: Source  \
user_id                                                                       
76561198096443555                         -0.666667                     NaN   
ImWinKo                                   -1.333333                     NaN   
ShevilleWarhand                                 NaN                     NaN   
WCM03                                      0.875000                   0.875   
tarjla                                     0.285714                     NaN   

item_name              DayZ  Five Nights at Freddy's  Garry's Mod  \
user_id                                                             
76561198096443555 -0.666667                      NaN          NaN   
ImWinKo            0.666667                      NaN          NaN   
ShevilleWarhand         NaN                 0.333333          NaN   
WCM03                   NaN                      NaN          NaN   
tarjla                  NaN                      NaN     0.285714   

item_name          Heroes &amp; Generals  Middle-earth™: Shadow of Mordor™  \
user_id                                                                      
76561198096443555               1.333333                          0.333333   
ImWinKo                         0.666667                               NaN   
ShevilleWarhand                      NaN                               NaN   
WCM03                           0.875000                               NaN   
tarjla                               NaN                               NaN   

item_name          No More Room in Hell  ...  \
user_id                                  ...   
76561198096443555                   NaN  ...   
ImWinKo                             NaN  ...   
ShevilleWarhand                     NaN  ...   
WCM03                            -1.125  ...   
tarjla                              NaN  ...   

item_name          Red Orchestra 2: Heroes of Stalingrad with Rising Storm  \
user_id                                                                      
76561198096443555                                                NaN         
ImWinKo                                                          NaN         
ShevilleWarhand                                                  NaN         
WCM03                                                         -0.125         
tarjla                                                           NaN         

item_name              Rust  Sid Meier's Civilization® V  Starbound  \
user_id                                                               
76561198096443555 -0.666667                          NaN        NaN   
ImWinKo            0.666667                          NaN        NaN   
ShevilleWarhand         NaN                          NaN        NaN   
WCM03             -1.125000                        0.875        NaN   
tarjla             0.285714                          NaN   0.285714   

item_name          Team Fortress 2  Terraria  The Forest     Trove  Unturned  \
user_id                                                                        
76561198096443555              NaN       NaN         NaN       NaN       NaN   
ImWinKo                        NaN       NaN   -1.333333       NaN       NaN   
ShevilleWarhand           0.333333       NaN    0.333333 -1.666667       NaN   
WCM03                          NaN       NaN   -1.125000       NaN       NaN   
tarjla                         NaN  0.285714   -1.714286       NaN  0.285714   

item_name            Verdun  
user_id       

In [40]:
# Eliminamos los juegos jugados por el usuario elegido
juegos_de_usuarios_similares.drop(juegos_jugados_por_usuario_elegido.columns,axis=1, inplace=True, errors='ignore')

# Y veamos la tabla resultante
juegos_de_usuarios_similares

item_name            Arma 3  Call of Duty®: Black Ops III  \
user_id                                                     
76561198096443555       NaN                           NaN   
ImWinKo            0.666667                           NaN   
ShevilleWarhand         NaN                      0.333333   
WCM03                   NaN                           NaN   
tarjla                  NaN                           NaN   

item_name          Counter-Strike: Global Offensive  Counter-Strike: Source  \
user_id                                                                       
76561198096443555                         -0.666667                     NaN   
ImWinKo                                   -1.333333                     NaN   
ShevilleWarhand                                 NaN                     NaN   
WCM03                                      0.875000                   0.875   
tarjla                                     0.285714                     NaN   

item_name              DayZ  Five Nights at Freddy's  \
user_id                                                
76561198096443555 -0.666667                      NaN   
ImWinKo            0.666667                      NaN   
ShevilleWarhand         NaN                 0.333333   
WCM03                   NaN                      NaN   
tarjla                  NaN                      NaN   

item_name          Middle-earth™: Shadow of Mordor™  No More Room in Hell  \
user_id                                                                     
76561198096443555                          0.333333                   NaN   
ImWinKo                                         NaN                   NaN   
ShevilleWarhand                                 NaN                   NaN   
WCM03                                           NaN                -1.125   
tarjla                                          NaN                   NaN   

item_name          Realm of the Mad God  \
user_id                                   
76561198096443555                   NaN   
ImWinKo                             NaN   
ShevilleWarhand                0.333333   
WCM03                               NaN   
tarjla                              NaN   

item_name          Red Orchestra 2: Heroes of Stalingrad with Rising Storm  \
user_id                                                                      
76561198096443555                                                NaN         
ImWinKo                                                          NaN         
ShevilleWarhand                                                  NaN         
WCM03                                                         -0.125         
tarjla                                                           NaN         

item_name              Rust  Sid Meier's Civilization® V  Starbound  \
user_id                                                               
76561198096443555 -0.666667                          NaN        NaN   
ImWinKo            0.666667                          NaN        NaN   
ShevilleWarhand         NaN                          NaN        NaN   
WCM03             -1.125000                        0.875        NaN   
tarjla             0.285714                          NaN   0.285714   

item_name          Team Fortress 2  Terraria  Unturned    Verdun  
user_id                                                           
76561198096443555              NaN       NaN       NaN  0.333333  
ImWinKo                        NaN       NaN       NaN       NaN  
ShevilleWarhand           0.333333       NaN       NaN       NaN  
WCM03                          NaN       NaN       NaN       NaN  
tarjla                         NaN  0.285714  0.285714       NaN

In [41]:
len(juegos_de_usuarios_similares.columns) # 17 juegos posibles para recomendar

17

In [42]:
# Creamos un diccionario para guardar las puntuaciones promedio de los juegos
puntaje_promedio = {}

# Iteramos sobre los juegos de los usuarios similares (que el usuario elegido no jugó)
for juego in juegos_de_usuarios_similares.columns:
  # Obtenemos el puntaje para el juego sobre el cuál se esté iterando
  puntaje_juego = juegos_de_usuarios_similares[juego]

  # Create a variable to store the score Variable para guardar el puntaje total del juego
  total = 0

  # Variable para contar la cantidad de puntajes que tiene el juego dado
  cantidad = 0

  # Iteramos entre los usuarios similares
  for usuario in usuarios_similares.index:

    # Si el juego fué punteado por el usuario
    if pd.isna(puntaje_juego[usuario]) == False:

      # puntaje_parcial es la multiplicación entre el puntaje entre ambos usuarios (el elegido y el iterado) y entre el puntaje que el usuario iterado le dió al juego en cuestión
      puntaje_parcial = usuarios_similares[usuario] * puntaje_juego[usuario]

      # Agregamos el puntaje parcial al total previamente creado
      total += puntaje_parcial

      # Agregamos 1 a la cantidad
      cantidad += 1
  
  # Luego de iterar sobre cada usuario, obtenemos el puntaje promedio para el juego iterado. El juego se almacena cómo clave dentro del diccionario y el valor es el promedio
  puntaje_promedio[juego] = total / cantidad

# Convertimos el diccionario a un dataframe
item_score = pd.DataFrame(puntaje_promedio.items(), columns=['item_name', 'item_score'])

# Ordenamos los juegos por su puntaje
ranked_item_score = item_score.sort_values(by='item_score', ascending=False)

# Mostramos el top m juegos
m = 5
ranked_item_score.head(m)

item_name  item_score
3        Counter-Strike: Source    0.333712
11  Sid Meier's Civilization® V    0.333712
0                        Arma 3    0.288675
15                     Unturned    0.132260
14                     Terraria    0.132260

Doy por finalizado la creación del sistema de recomendación, en esta misma carpeta voy a crear otro archivo en dónde desarrolle el sistema de recomendación dentro de una función